cmd/engine/executor.go

```diff
@@ -0,0 +1,1053 @@
+
+	defer executor.MountStubsCleaner(ctx, rootFSPath, mounts, meta.RemoveMountStubsRecursive)()
+
+	setupCACerts := true
+	return nil, w.run(
+		ctx,
+		id,
+		root,
+		mounts,
+		rootFSPath,
+		process,
+		namespace,
+		started,
+		setupCACerts,
+	)
+}
+
+func (w *runcExecutor) run(
+	ctx context.Context,
+	id string,
+	root executor.Mount,
+	mounts []executor.Mount,
+	rootFSPath string,
+	process executor.ProcessInfo,
+	namespace network.Namespace,
+	started chan<- struct{},
+	installCACerts bool,
+) (err error) {
+	startedOnce := sync.Once{}
+	done := make(chan error, 1)
+	w.mu.Lock()
+	w.running[id] = done
+	w.mu.Unlock()
+	defer func() {
+		w.mu.Lock()
+		delete(w.running, id)
+		w.mu.Unlock()
+		done <- err
+		close(done)
+		if started != nil {
+			startedOnce.Do(func() {
+				close(started)
+			})
+		}
+	}()
+
+	bundle := filepath.Join(w.root, id)
+	if err := os.Mkdir(bundle, 0o711); err != nil {
+		return err
+	}
+	defer os.RemoveAll(bundle)
+
+	uid, gid, sgids, err := oci.GetUser(rootFSPath, process.Meta.User)
+	if err != nil {
+		return err
+	}
+
+	resolvConf, err := oci.GetResolvConf(ctx, w.root, w.idmap, w.dns, process.Meta.NetMode)
+	if err != nil {
+		return err
+	}
+
+	hostsFile, clean, err := oci.GetHostsFile(ctx, w.root, process.Meta.ExtraHosts, w.idmap, process.Meta.Hostname)
+	if err != nil {
+		return err
+	}
+	if clean != nil {
+		defer clean()
+	}
+
+	configPath := filepath.Join(bundle, "config.json")
+	f, err := os.Create(configPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	opts := []containerdoci.SpecOpts{oci.WithUIDGID(uid, gid, sgids)}
+
+	if process.Meta.ReadonlyRootFS {
+		opts = append(opts, containerdoci.WithRootFSReadonly())
+	}
+
+	identity := idtools.Identity{
+		UID: int(uid),
+		GID: int(gid),
+	}
+	if w.idmap != nil {
+		identity, err = w.idmap.ToHost(identity)
+		if err != nil {
+			return err
+		}
+	}
+
+	spec, cleanup, err := oci.GenerateSpec(
+		ctx,
+		process.Meta,
+		mounts,
+		id,
+		resolvConf,
+		hostsFile,
+		namespace,
+		w.cgroupParent,
+		w.processMode,
+		w.idmap,
+		w.apparmorProfile,
+		w.selinux,
+		w.tracingSocket,
+		opts...,
+	)
+	if err != nil {
+		return err
+	}
+	defer cleanup()
+
+	spec.Root.Path = rootFSPath
+	if root.Readonly {
+		spec.Root.Readonly = true
+	}
+
+	newp, err := fs.RootPath(rootFSPath, process.Meta.Cwd)
+	if err != nil {
+		return errors.Wrapf(err, "working dir %s points to invalid target", newp)
+	}
+	if _, err := os.Stat(newp); err != nil {
+		if err := idtools.MkdirAllAndChown(newp, 0o755, identity); err != nil {
+			return errors.Wrapf(err, "failed to create working directory %s", newp)
+		}
+	}
+
+	spec.Process.Terminal = process.Meta.Tty
+	spec.Process.OOMScoreAdj = w.oomScoreAdj
+
+	if err := json.NewEncoder(f).Encode(spec); err != nil {
+		return err
+	}
+	f.Close()
+
+	bklog.G(ctx).Debugf("> creating %s %v", id, process.Meta.Args)
+	defer bklog.G(ctx).Debugf("> container done %s %v", id, process.Meta.Args)
+
+	if installCACerts {
```

Comment:
Just a note for reviewers, this code is mostly the same as upstream's runc executor except for this section that handles CA cert installation and some other minor re-arrangement/cleanup of irrelevant codepaths.